src/core/sys/linux/perf_event.d

```diff
@@ -0,0 +1,2468 @@
+ *        { u64    id;           } && PERF_FORMAT_ID
+ *      }        cntr[nr];
+ *    } && PERF_FORMAT_GROUP
+ * };
+ * ---
+ */
+enum perf_event_read_format
+{
+        ///
+        PERF_FORMAT_TOTAL_TIME_ENABLED = 1U << 0,
+        ///
+        PERF_FORMAT_TOTAL_TIME_RUNNING = 1U << 1,
+        ///
+        PERF_FORMAT_ID = 1U << 2,
+        ///
+        PERF_FORMAT_GROUP = 1U << 3,
+        PERF_FORMAT_MAX = 1U << 4 /** non-ABI */
+}
+
+enum PERF_ATTR_SIZE_VER0 = 64; /** sizeof first published struct */
+enum PERF_ATTR_SIZE_VER1 = 72; /** add: config2 */
+enum PERF_ATTR_SIZE_VER2 = 80; /** add: branch_sample_type */
+enum PERF_ATTR_SIZE_VER3 = 96; /** add: sample_regs_user */
+/* add: sample_stack_user */
+enum PERF_ATTR_SIZE_VER4 = 104; /** add: sample_regs_intr */
+enum PERF_ATTR_SIZE_VER5 = 112; /** add: aux_watermark */
+
+/**
+ * Hardware event_id to monitor via a performance monitoring event:
+ *
+ * @sample_max_stack: Max number of frame pointers in a callchain,
+ *              should be < /proc/sys/kernel/perf_event_max_stack
+ */
+struct perf_event_attr
+{
+        /**
+        *Major type: hardware/software/tracepoint/etc.
+        */
+        uint type;
+
+        /**
+        * Size of the attr structure, for fwd/bwd compat.
+        */
+        uint size;
+
+        /**
+        * Type specific configuration information.
+        */
+        ulong config;
+        ///
+        union
+        {
+                ///
+                ulong sample_period;
+                ///
+                ulong sample_freq;
+        }
+        ///
+        ulong sample_type;
+        ///
+        ulong read_format;
+
+        // mixin(bitfields!(
+        //     ulong, "disabled", 1,
+        //     ulong, "inherit", 1,
+        //     ulong, "pinned", 1,
+        //     ulong, "exclusive", 1,
+        //     ulong, "exclude_user", 1,
+        //     ulong, "exclude_kernel", 1,
+        //     ulong, "exclude_hv", 1,
+        //     ulong, "exclude_idle", 1,
+        //     ulong, "mmap", 1,
+        //     ulong, "comm", 1,
+        //     ulong, "freq", 1,
+        //     ulong, "inherit_stat", 1,
+        //     ulong, "enable_on_exec", 1,
+        //     ulong, "task", 1,
+        //     ulong, "watermark", 1,
+        //     ulong, "precise_ip", 2,
+        //     ulong, "mmap_data", 1,
+        //     ulong, "sample_id_all", 1,
+        //     ulong, "exclude_host", 1,
+        //     ulong, "exclude_guest", 1,
+        //     ulong, "exclude_callchain_kernel", 1,
+        //     ulong, "exclude_callchain_user", 1,
+        //     ulong, "mmap2", 1,
+        //     ulong, "comm_exec", 1,
+        //     ulong, "use_clockid", 1,
+        //     ulong, "context_switch", 1,
+        //     ulong, "write_backward", 1,
+        //     ulong, "namespaces", 1,
+        //     ulong, "__reserved_1", 35));
+        private ulong perf_event_attr_bitmanip;
+        ///
+        @property ulong disabled() @safe pure nothrow @nogc const
```

Comment:
There's functions here and no unittests?  How can we vet that this actually works?